File: App/Sources/Environment/AppEnvironment.swift
import Foundation
import TeamTasksCore
import TeamTasksMocks
import TeamTasksSupabase

/// The backend and platform services chosen once at launch.
///
/// - `-uiTestMockBackend` (UI tests, demos): in-memory backend of TeamTasksMocks with the French demo data,
///   in the state given by `-mockScenario <signedOut|populated|emptyGroups>` (default `signedOut`). Optional:
///   `-mockLatencyMs <n>` (artificial latency of every call) and
///   `-mockNotifications <notDetermined|denied|authorized>` (initial permission of the in-app notification fake).
///   Nothing is persisted and no system prompt is ever shown.
/// - Otherwise: the Supabase project of Info.plist (`SupabaseScheme`, `SupabaseHost`, `SupabasePublishableKey`),
///   local notifications through `UNUserNotificationCenter`, settings in `UserDefaults`.
struct AppEnvironment: Sendable {
    enum Backend: Sendable, Equatable {
        case mock(MockScenario)
        case supabase(URL)
    }

    static let mockBackendArgument = "-uiTestMockBackend"
    static let mockLatencyArgument = "-mockLatencyMs"
    static let mockNotificationsArgument = "-mockNotifications"

    let backend: Backend
    let services: AppServices
    let platform: PlatformServices

    /// True with the in-memory backend (UI tests): no background refresh, no system permission prompt.
    var isMock: Bool {
        if case .mock = backend { return true }
        return false
    }

    /// The environment for these launch arguments and this bundle's Info.plist.
    static func resolve(
        arguments: [String] = ProcessInfo.processInfo.arguments,
        bundle: Bundle = .main
    ) -> Result<AppEnvironment, ConfigurationIssue> {
        if arguments.contains(mockBackendArgument) {
            return .success(mock(arguments: arguments))
        }
        return SupabaseSettings.configuration(from: bundle).map { configuration in
            AppEnvironment(
                backend: .supabase(configuration.url),
                services: SupabaseBackend.makeServices(configuration: configuration),
                platform: devicePlatform()
            )
        }
    }

    // MARK: - Mock backend

    static func mock(arguments: [String]) -> AppEnvironment {
        let scenario = MockEnvironment.scenario(fromLaunchArguments: arguments) ?? .signedOut
        let latencyMilliseconds = integer(after: mockLatencyArgument, in: arguments) ?? 0
        let mock = MockEnvironment.make(scenario: scenario, latency: .milliseconds(max(0, latencyMilliseconds)))
        let platform = PlatformServices(
            notifications: InAppNotificationScheduler(status: notificationStatus(in: arguments)),
            store: InMemoryKeyValueStore(),
            now: { Date() },
            // The demo due dates are wall-clock times in Europe/Paris: show them as such whatever the
            // simulator's time zone, so screenshots are stable.
            calendar: DemoData.calendar
        )
        return AppEnvironment(backend: .mock(scenario), services: mock.services, platform: platform)
    }

    private static func notificationStatus(in arguments: [String]) -> NotificationAuthorization {
        switch string(after: mockNotificationsArgument, in: arguments) {
        case "authorized": .authorized
        case "denied": .denied
        default: .notDetermined
        }
    }

    private static func string(after flag: String, in arguments: [String]) -> String? {
        guard let index = arguments.firstIndex(of: flag), arguments.indices.contains(index + 1) else { return nil }
        return arguments[index + 1]
    }

    private static func integer(after flag: String, in arguments: [String]) -> Int? {
        string(after: flag, in: arguments).flatMap { Int($0) }
    }

    // MARK: - Device

    /// Platform services of a real device (docs/CONTRACTS.md §7).
    static func devicePlatform() -> PlatformServices {
        PlatformServices(
            notifications: UserNotificationScheduler(),
            store: UserDefaultsKeyValueStore(),
            now: { Date() },
            calendar: AppCalendar.french
        )
    }
}

File: android/app/src/main/kotlin/io/github/notkanaa/equipe/config/AppEnvironment.kt
package io.github.notkanaa.equipe.config

import android.content.Context
import android.content.Intent
import io.github.notkanaa.equipe.BuildConfig
import io.github.notkanaa.equipe.core.AppCalendar
import io.github.notkanaa.equipe.core.AppServices
import io.github.notkanaa.equipe.core.InMemoryKeyValueStore
import io.github.notkanaa.equipe.core.NotificationAuthorization
import io.github.notkanaa.equipe.core.PlatformServices
import io.github.notkanaa.equipe.mocks.DemoData
import io.github.notkanaa.equipe.mocks.MockEnvironment
import io.github.notkanaa.equipe.mocks.MockScenario
import io.github.notkanaa.equipe.platform.AndroidNotificationScheduler
import io.github.notkanaa.equipe.platform.InAppNotificationScheduler
import io.github.notkanaa.equipe.platform.SharedPreferencesKeyValueStore
import io.github.notkanaa.equipe.platform.SharedPreferencesSessionStorage
import io.github.notkanaa.equipe.supabase.SupabaseBackend
import io.github.notkanaa.equipe.supabase.SupabaseConfiguration
import java.time.Instant
import java.time.ZoneId
import kotlin.time.Duration.Companion.milliseconds

/**
 * A debug launch on the in-memory backend of :mocks (UI tests, demos, screenshots): the launch intent's extras
 * (`adb shell am start -n io.github.notkanaa.equipe/.MainActivity --es mockScenario populated`), or the debug build's
 * permanent scenario (`-Pequipe.mockScenario=populated`). Release builds ignore both.
 *
 * - [EXTRA_SCENARIO] `mockScenario`: `signedOut`, `populated` or `emptyGroups` (iOS `-mockScenario`);
 * - [EXTRA_NOTIFICATIONS] `mockNotifications` (optional): `notDetermined` (default), `denied` or `authorized`, the
 *   initial permission of the in-app notification fake (no system prompt, no real notification);
 * - [EXTRA_LATENCY_MS] `mockLatencyMs` (optional, string or int): artificial latency of every backend call.
 */
data class MockLaunchOptions(
    val scenario: MockScenario,
    val notifications: NotificationAuthorization = NotificationAuthorization.NOT_DETERMINED,
    val latencyMillis: Long = 0,
) {
    companion object {
        const val EXTRA_SCENARIO: String = "mockScenario"
        const val EXTRA_NOTIFICATIONS: String = "mockNotifications"
        const val EXTRA_LATENCY_MS: String = "mockLatencyMs"

        /** The options of a launch intent; null without a valid `mockScenario` extra, and always in release builds. */
        fun fromIntent(intent: Intent?): MockLaunchOptions? {
            if (!BuildConfig.MOCK_BACKEND_AVAILABLE || intent == null) return null
            val scenario = intent.getStringExtra(EXTRA_SCENARIO)?.let(MockScenario::fromRawValue) ?: return null
            val latency = intent.getStringExtra(EXTRA_LATENCY_MS)?.trim()?.toLongOrNull()
                ?: intent.getIntExtra(EXTRA_LATENCY_MS, 0).toLong()
            return MockLaunchOptions(
                scenario = scenario,
                notifications = notificationStatus(intent.getStringExtra(EXTRA_NOTIFICATIONS)),
                latencyMillis = latency.coerceAtLeast(0),
            )
        }

        /** The debug build's permanent scenario (`-Pequipe.mockScenario=…`); null otherwise and in release builds. */
        fun fromBuildConfig(): MockLaunchOptions? {
            if (!BuildConfig.MOCK_BACKEND_AVAILABLE) return null
            val scenario = MockScenario.fromRawValue(BuildConfig.MOCK_SCENARIO) ?: return null
            return MockLaunchOptions(scenario)
        }

        private fun notificationStatus(rawValue: String?): NotificationAuthorization = when (rawValue) {
            "authorized" -> NotificationAuthorization.AUTHORIZED
            "denied" -> NotificationAuthorization.DENIED
            else -> NotificationAuthorization.NOT_DETERMINED
        }
    }
}

/** The backend and platform services chosen once per launch (port of the iOS `AppEnvironment`). */
class AppEnvironment private constructor(
    val backend: Backend,
    val services: AppServices,
    val platform: PlatformServices,
) {
    sealed interface Backend {
        /** In-memory backend of :mocks (debug builds only). */
        data class Mock(val options: MockLaunchOptions) : Backend

        /** The Supabase project at [url]. */
        data class Supabase(val url: String) : Backend
    }

    /** True with the in-memory backend: no background refresh, no system permission prompt, nothing persisted. */
    val isMock: Boolean get() = backend is Backend.Mock

    companion object {
        /**
         * The Supabase backend: the Auth session persisted in SharedPreferences (excluded from backups), local
         * notifications through NotificationManager + AlarmManager, settings in SharedPreferences, the device's time
         * zone. Call it once per process ([SupabaseBackend.makeServices] builds a new client session).
         */
        fun supabase(context: Context, configuration: SupabaseConfiguration): AppEnvironment {
            val appContext = context.applicationContext
            return AppEnvironment(
                backend = Backend.Supabase(configuration.url),
                services = SupabaseBackend.makeServices(configuration, SharedPreferencesSessionStorage(appContext)),
                platform = PlatformServices(
                    notifications = AndroidNotificationScheduler(appContext),
                    store = SharedPreferencesKeyValueStore(appContext),
                    now = { Instant.now() },
                    calendar = AppCalendar.frenchGregorian(ZoneId.systemDefault()),
                ),
            )
        }

        /** The in-memory backend in the state of [options] (fresh demo data every time). */
        fun mock(options: MockLaunchOptions): AppEnvironment {
            val environment = MockEnvironment.make(options.scenario, latency = options.latencyMillis.milliseconds)
            return AppEnvironment(
                backend = Backend.Mock(options),
                services = environment.services,
                platform = PlatformServices(
                    notifications = InAppNotificationScheduler(options.notifications),
                    store = InMemoryKeyValueStore(),
                    now = { Instant.now() },
                    // The demo due dates are wall-clock times in Europe/Paris: shown as such whatever the device's
                    // time zone, so that screenshots are stable (like iOS).
                    calendar = DemoData.calendar,
                ),
            )
        }
    }
}
